### DISCUSSION TITLE
canDo() - Testing in advance when emitting and event would result in a transition or not.

### DISCUSSION THREAD

**Discussion by szabeszg** [shuckster/statebot#2] [Ideas]
Hello,

Thanks for sharing your state machine project, I like it a lot, however, I'm missing at least one feature: in (for example) Dave Stewart's "proof of concept" FSM it is possible to query the transition chart to see if an action is available from the current state, meaning if emitting a particular event would result in a transition or not.

see: https://github.com/davestewart/javascript-state-machine/blob/master/docs/api/statemachine.md#canDo

This would be particularly useful to update GUI elements based on the result of such a function call. When emitting an event would  result in no state change then I need to disbale buttons in advance, for example.

**Comment by shuckster** [comment]
Thank you for your feedback and observation.

Statebot has two methods that deal with upcoming states:

- [#canTransitionTo](https://zansh.in/statebot/#statebotfsmcantransitionto)
- [#stateAvailableFromHere](https://zansh.in/statebot/#statebotfsmstatesavailablefromhere)

Unfortunately I don't think they quite line-up with the specification you linked-to because `canDo` takes the current action/event. But Statebot charts specify only state-transitions, and this is its only internal mapping.

Events can be specified in [#performTransitions](https://zansh.in/statebot/#statebotfsmperformtransitions) for example, but these are applied more like `addEventHandler`, so there is no corresponding dictionary against the states they might cause a transition to.

However, I wouldn't say that the scenario you outline is impossible to implement with Statebot, but perhaps it may be tricky to make it elegant within a single machine.

Take this contrived example:

```ini
[entering-email] -> email-valid | email-invalid
  email-invalid -> entering-email
  email-valid -> [entering-password]

[entering-password] -> password-valid | password-invalid
  password-invalid -> entering-password
  password-valid -> [form-completed]

[form-completed]
```

Using the above chart, perhaps an email-form might look something like:

```jsx
const emailHasValidFormat = machine.inState('email-valid')
const submittedInvalidEmail = machine.history().includes('email-invalid')

return (
  <form>
    <input type="text" name="email" />

    {submittedInvalidEmail && (
      <p>Your email is invalid. Please try again.</p>
    )}

    <submit disabled={!emailHasValidFormat}>Submit</submit>
  </form>
)
```

As you can see, only [#inState](https://zansh.in/statebot/#statebotfsminstate) was required in this case.

I can imagine such examples could be solved with more than one Statebot, which might keep each one simple rather than try to have an exhaustive chart as I've tried to attempt in the above thrown-together example.

 Happy to hear more on it though. Do you have a more complete example?

**Reply by szabeszg** [reply]
Thank you for your response!

> Happy to hear more on it though. Do you have a more complete example?

If I'm not mistaken, in your example above the logic to enable/disable the button is based on the current state. The message to the user is based on a previous state. What I don't get is how this example solves the issue of figuring out the existence of a next "possible state" the FMS gets into, should the program emit a certain event.

My example is simple, I have uploaded it to here: https://szabesz.hu/_dl/c3/statebot/index.html
This is just dead simple sequence of on/off states ("pretend traffic lights") with buttons for manually advancing forwards and backwards.

I am working with [Construct 3](https://www.construct.net/en), trying to separate state management from the "view". IN my Construct project, the JavaScript bit coded by me should manage the sates while Construct's game engine should be responsible for managing the view.

```JavaScript
const initState = "green";
const next = "green -> amber -> red";
const back = "red -> amber -> green";

fsmNextBack = Statebot('fsmNextBack', {
  chart: `
	${next}
	${back}
  `,
  logLevel: 3,
  startIn: initState,
})

fsmNextBack.performTransitions({
	'green -> amber -> red': { on: 'next' },
	'red -> amber -> green': { on: 'back' },
})

fsmNextBack.onEvent('next', () => {redrawTrafficLights('next');})
fsmNextBack.onEvent('back', () => {redrawTrafficLights('back');})

function redrawTrafficLights(event) {
	fsmNextBack.inState({
		'red': () => {
			//console.log('draw red light, enable (true) or disbale (false) Next & Back buttons');
			runtime.callFunction("updateButtons", false, true);
			runtime.callFunction("updateLights");
		},
		'amber': () => {
			//console.log('draw amber light, enable (true) or disbale (false) Next & Back buttons');
			runtime.callFunction("updateButtons", true, true);
			runtime.callFunction("updateLights");
		},
		'green': () => {
			//console.log('draw green light, enable (true) or disbale (false) Next & Back buttons');
			runtime.callFunction("updateButtons", true, false);
			runtime.callFunction("updateLights");
		},
	})
}

redrawTrafficLights();
```

Currently, I just hard coded in what states the next/back buttons must be in, by passing true/false to the updateButtons function. The updateLights function checks the current state of the state machine so there is no need to pass along anything in that case.

In this simple example project of mine it is not a big deal to "hard code" the required button states this way, but using a state machine I would like to be able to do it based on the state chart instead, especially in more complex projects.

Dave Stewart's similar example solves this issue by using his machines's canDo method:
https://statemachine.davestewart.io/html/examples/flows/sequential.html

I understand that "there is no corresponding dictionary against the states they might cause a transition to" but would it be possible to implement it in a future version perhaps?

Side note: I spent a lot of time trying to find a JavaScript FSM library and XState is just "too much" for my needs as I would not need most its features (it would be overkill for my Construct projects). Looks like Statebot is the only lightweight fsm implementation that fits my projects. Another one could be Dave's state machine but that is just a "[proof of concept](https://davestewart.co.uk/projects/open-source/state-machine/)" project.

**Reply by shuckster** [reply]
Thank you for the detailed reply and the example.

I can see now why it would be useful to predict the next state given an event.

Since Statebot uses an event-emitting library internally, it seems reasonable that it should notice when events are configured, and if emitting them would result in a transition.

I propose a couple of new methods:

## `stateAfterEmitting`

```js
stateAfterEmitting(eventName)
// -> string
```

Peeks at what the next state will be after emitting `eventName`. If this will not result in a *new* state, the `currentState()` will be returned.

- Will `throw` if: Not enough args
- Will log an `error` if: More than one transition is possible after emitting a single event (likely config issue)
- Will log a `warning` if: The specified event is not being listened-for by this machine

## `inStateAfterEmitting`

```js
inStateAfterEmitting(eventName, expectedState)
// -> true | false
```

Will throw/log on the same conditions as `stateAfterEmitting`.

How do these look?

**Reply by szabeszg** [reply]
They look good, thanks. However, while logging errors/warning is useful for debugging, such information might also be useful for the running application.

Say, I want to write an "unfolding story" game, in which the player can decide what path to choose next. So the player is asked:
_"Do you go north or south? Before you decide, your instinct says that it is more dangerous to go to north than south."_

The last statement could be based on stateAfterEmitting() returning the number of possible new states. Probing one transition would result in less possible states (going south) than probing the other (going north). The more possible states ahead, the more uncertain future awaits you... :)

Second, I guess, if stateAfterEmitting() returns the current state to indicate that no state change will occur, then it returns the next state if there is only one possible state to transition to. Still, it could return a "special" string (say) "__multiple" to indicate the case of more than one possible states, or – if that is possible to implement – the exact number of such states, or even better: an array of the possible state names. (I don't know if you rule out returning mixed types or not.)

As for the "specified event is not being listened-for by this machine" case, being able to test it runtime would also be useful.

In another game example, the player receives a message like:
_"You tried to hit a good fairy. Shame on you!"_
And this message is based on the fact that no such action can be performed in the game but trying to perform a hit action on a fairy would not be a good thing to do anyway.

The latter example could be implemented with its own state (eg. a "fairy_was_hit" state) and the application could respond to such a state change, but if a state exists in the Chart only for "not being able to actually perform something", than the indication of an "event is not being listened-for" could simplify the Chart in certain scenarios.

BTW, regarding what the methods are called, stateAfterEmitting() and inStateAfterEmitting() both sound the same to me (even though I get the difference). Maybe instead of inStateAfterEmitting() something like canTransitionToByEmitting(eventName, expectedStates) would be better, which is similar to canTransitionTo (states).

I hope all this makes sense to you.

**Reply by shuckster** [reply]
I'll try and summarise what you've said to see if I've got the picture straight:

`stateAfterEmitting` should:

1. ... have the possibility to return multiple states, either as a special value or as an array of values.

2. ... indicate if a state-transition will not happen.

3. ... indicate if an event is not registered.

As for the name of `inStateAfterEmitting`, I agree that `canTransitionToByEmitting` sounds more consistent, so thank you for that suggestion.

To address points 1-3:

1. Does [#statesAvailableFromHere](https://zansh.in/statebot/#statebotfsmstatesavailablefromhere) not already cover this? If you wish to know *all* potential states from the current state, you don't need to involve an event.

    However, it does not make sense to me to return more than one state if you are asking what will happen _if you emit an event_. For me, emitting an event should always result in either:

    - a single transition to the next state
    - nothing

    If multiple states are possible from a single event, then this would indicate that the machine has not been well configured. I mean, it's *true* that an event emitted many times can cause a machine to visit many states *over time*. But a single event should result in a single transition, which I believe is what `canDo`/`stateAfterEmitting` should indicate.

2. Yes, it might indeed by useful to know if a state-transition will not happen, but I think this could be satisfactorily achieved like this:

    ```js
    const stuck = fsm.stateAfterEmitting('go-north') === fsm.currentState()
    ```

    Alternatively, `undefined` could be returned therefore allowing the user to identify the next-state like this:

    ```js
    const nextState = fsm.stateAfterEmitting('go-north') ?? fsm.currentState()
    const stuck = fsm.stateAfterEmitting('go-north') === undefined
    ```

    For me, I think I prefer the first option just because "next-state equals current-state" makes semantic sense if you're trying to figure out if the machine is "stuck". Comparing to `undefined` is more ambiguous.

    But perhaps there's room for a more extended syntax like `inState` has?

    Something like this:

    ```js
    const upNext = fsm.peek('go-north', {
      'entrance': () => 'Ahead is a place to hang your coat',
      'courtyard': () => 'Through the windows you see overgrown monuments',
      'paddock': () => 'A stream of water burbles ahead',
      'woods': () => 'The trees are tall and thin and many',
      undefined: () => 'You cannot go north',
    })
    ```

    Hopefully this makes sense? (By the way, `peek` is an alternative name for `stateAfterEmitting`. I think I prefer it, especially as it doesn't seem to clash so badly with the extended syntax above.)

    Actually, this syntax could allow `undefined` to be returned like this:

    ```js
    const nextState = fsm.peek('go-north', {
      undefined: undefined
    })
    // Force peek() to return `undefined` instead of the current-state
    // if no transition is possible. (States with no key in the object
    // will be just be returned as-is.)
    ```

    I don't like it, but it would be possible. :)

3. For the case where an event is not registered, I'm not sure if this one should be behaviourally any different from the previous case.

    I'm just trying to think about what's important to know if a transition isn't going to happen...

    If being stuck can already be determined by:

    ```js
    const stuck = fsm.peek('go-north') === fsm.currentState()
    ```

    Then I wonder about the value of being more granular than this.

    I think logging would still be useful. It's good to know if a transition didn't happen because an event isn't registered. But as for what to return, I am thinking that for the general case (ie; not just games) it may be better to always return a state.

That's my thinking so far. Let me know if I understood correctly.

**Reply by szabeszg** [reply]
Hi, regarding points 1-3

1. I was thinking along the line of "conditional transitions" so I misinterpreted why you proposed the "error" message, and now I get what you meant by "likely config issue". So yes, there is no need to implement "runtime error checking" like that, I guess.

2. I like the "peek" version, it is a lot shorter and an even more descriptive than stateAfterEmitting, as "state after emitting" could also be interpreted by the coder as the state after an event has actually been emitted, but peek sounds more like testing beforehand.

I also like your upNext example, looks simple but powerful.

3. When I know what event I tested for emitting and I also know that emitting that event would not result in a state change, then I can of course rely on that info to output something I described in my previous post. I was just thinking that if the method already detects such a scenario for outputting a warning message, then it might be also useful to use it in the program itself as a "shorthand technique". Also, such an application state might not mean it is an issue with the code when it is handled gracefully. Well, eat least if I understand correctly what you propose.

**Reply by shuckster** [reply]
Thank you for the further feedback.

I believe the combination of `peek` and `canTransitionToAfterEmitting` should be able to cover the use-cases you bring up.

So with the following config:

```js
const machine = Statebot('peek-a-boo', {
  chart: `
    idle -> running
  `
})

machine.performTransitions({
  'idle -> running': {
    on: 'start'
  }
})
```

We could do this:

```js
1: machine.peek('start')
// "running"

2: machine.peek('non-existent-event')
// "idle"
// -> ie; the current-state

3: machine.canTransitionToAfterEmitting('non-existent-event', 'running')
// false
// -> no transition will happen
```

So with `2:`, no transition will happen and we have to detect it in another step (ie; by comparing the result to `currentState()`.) But with `3:` we don't need to take that step.

You make a good case for returning a different result depending on whether or not a warning is emitted, but I'm not sure it convinces me. I think I'd prefer a dedicated method for that, perhaps called `isEventRegistered`? But since FSM config is specified up-front I wonder about the utility.

Finally, after looking at `canTransitionToAfterEmitting` I find it a bit long, so have come up with these alternatives:

```js
1: machine.canTransitionTo('running', { afterEmitting: 'start' })

2: machine.afterEmitting('start', { willTransitionTo: 'running' })
```

I kind of prefer the first one. It's a bit annoying to overload an existing method, but maybe it's worth it?

**Reply by szabeszg** [reply]
Hello, I do like your proposal of `peek` and `canTransitionTo` or `afterEmitting`, thanks.

Regarding  `canTransitionTo` vs `afterEmitting`, I would vote for `canTransitionTo` even if it involves overloading, which is not a big deal when overloading is not overused (so to speak) like in this case, I think. I myself first looked at what `canTransitionTo` does when looking for a "`canDo`" method before starting this discussion.

BTW, are you still thinking of including this extended syntax as well, perhaps?
```JavaScript
const upNext = fsm.peek('go-north', {
  'entrance': () => 'Ahead is a place to hang your coat',
  'courtyard': () => 'Through the windows you see overgrown monuments',
  'paddock': () => 'A stream of water burbles ahead',
  'woods': () => 'The trees are tall and thin and many',
  undefined: () => 'You cannot go north',
})
```
In my point of view it would "just" be a "shortcut" but a useful one. However, such an option might rule out the possibility of sensible overloading?

**Reply by shuckster** [reply]
Good to hear you like the approach.

I'll get to work on both new methods and see if the overloading isn't too much of a pain for `canTransitionTo`.

I'll also try and get the `peek` "shorthand" working too.

**Comment by szabeszg** [comment]
Thank you so much! In the meantime, I refactored my current Construct project to make use of Statebot instead of a bunch of global variables and things are so much cleaner and shorter this way. When using the new methods, the inner workings of the interface state management will be even more streamlined.

I especially enjoy the verbose log messages as they point me to the right direction when things are just not yet working quite right.

In a day or two, I will also find the time to buy you some coffee, btw :)

**Reply by shuckster** [reply]
I managed to get the methods in as-discussed, including the [.peek()](https://zansh.in/statebot/#statebotfsmpeek) shorthand. Let me know if everything works as expected, and hope your refactoring goes well. There are tests for the new functionality, but I may not have caught everything.

Thank you so much for the kind words and support. It's really appreciated, not least because it has made the software better. 🙏

**Reply by shuckster** [reply]
[v2.9.1 released](https://www.npmjs.com/package/statebot). 📦

**Reply by szabeszg** [reply]
Hi, Thanks for your work and kind words. I'm about to update my current Construct project to use `peak()` and I will report back.

Just one more observation: as I do not use nmp, first I simply downloaded the zip from https://github.com/shuckster/statebot/releases/tag/v2.9.1
which results in 2.9.0 being downloaded instead of 2.9.1. Downloading the master brach leads to 2.9.1, but for some reason the tagged version does not.

**Reply by szabeszg** [reply]
As the state machine in my current project is tiny, all I had implemented in addition was only this:
```JavaScript
function can(eventToEmit) {
  return fsm.peek(eventToEmit) === fsm.currentState() ? 'disabled' : 'enabled';
}
```
and I also updated some method calls by using `can('restart')` and `can('accept')`

Works fine :) I will need to expand the state machine later on when I add some ajax email sending features, and I also have two, even more complex projects to produce after finishing the current one, so I will definitely do more than this, based on Statebot 2.9+.

**Reply by shuckster** [reply]
> which results in 2.9.0 being downloaded instead of 2.9.1

Oh, that's weird. I wonder if it has always happened like this? I use an automated deploy script, so I'll look into that thank you.

Glad to see you've made use of the new methods already! I look forward to hearing any further feedback or progress you'd like to share.

**Reply by shuckster** [reply]
The version mismatch might just have been a propagation issue. Saw the same thing today in a different repo.

Hopefully 2.9.2 (same as 2.9.1 but with updated `package.json`) is downloadable from Git now.

**Reply by szabeszg** [reply]
I see. Thanks for notifying me.
